adam_core/coordinates/residuals.py

```diff
@@ -1,161 +1,212 @@
+    p = np.zeros(N, dtype=np.float64)
+    chi2s = np.zeros(N, dtype=np.float64)
 
     # Caclulate the degrees of freedom for every coordinate
-    # Number of coordinate dimensions less the number of masked quantities
-    dof = observed.values.shape[1] - observed.values.mask.sum(axis=1)
+    # Number of coordinate dimensions less the number of quantities that are NaN
+    dof = D - np.sum(np.isnan(observed.values), axis=1)
 
+    # Calculate the array of residuals
     residuals = observed.values - predicted.values
-    # if isinstance(observed, SphericalCoordinates):
-    # residuals[:, 1] = np.where(residuals[:, 1] > 180., 360. - residuals[:, 1], residuals[:, 1])
-    # residual_ra *= np.cos(np.radians(dec_pred))
-    # residuals[:, 2] = np.where(residuals[:, 2] > 90., 360. - residuals[:, 2], residuals[:, 2])
-
-    if observed.covariances is not None:
-        chi2s = residuals**2 / observed.sigmas**2
-
-        d_list = []
-        p_list = []
-        for i, (observed_i, predicted_i) in enumerate(zip(observed, predicted)):
-            mask = observed_i.values.mask
-            dof_i = dof[i]
-
-            u = predicted_i.values[~mask].filled()
-            v = observed_i.values.compressed()
-            cov = observed_i.covariances.compressed().reshape(dof_i, dof_i)
-
-            d_i = mahalanobis(u, v, np.linalg.inv(cov))
-            p_i = 1 - chi2.cdf(d_i, dof_i)
-
-            d_list.append(d_i)
-            p_list.append(p_i)
-
-        d = np.array(d_list)
-        p = np.array(p_list)
-    else:
-        p = np.empty(len(observed))
-        p.fill(np.NaN)
-        d = np.empty(len(observed))
-        d.fill(np.NaN)
-
-        chi2s = np.ma.zeros(residuals.shape, dtype=float)
-        chi2s.fill(np.NaN)
-        chi2s.mask = residuals.mask
-        chi2s.fill_value = np.NaN
-
-    return Residuals(residuals, chi2s, dof, d, p, residual_names)
+
+    # Batch the coordinates and covariances into groups that have the same
+    # number of dimensions that have missing values (represented by NaNs)
+    (
+        batch_indices,
+        batch_dimensions,
+        batch_coords,
+        batch_covariances,
+    ) = batch_coords_and_covariances(observed.values, observed.covariances.to_matrix())
+
+    for i, (indices, dimensions, coords, covariances) in enumerate(
+        zip(batch_indices, batch_dimensions, batch_coords, batch_covariances)
+    ):
+        if not np.all(np.isnan(covariances)):
+            # Calculate the chi2 for each coordinate
+            chi2_values = calculate_chi2(residuals[indices], covariances)
+
+            # Calculate the probability for each coordinate
+            p[indices] = 1 - chi2.cdf(np.sqrt(chi2_values), dof[indices])
+
+            # Set the chi2 for each coordinate
+            chi2s[indices] = chi2_values
+
+        else:
+            # If the covariance matrix is all NaNs, then the chi2 is NaN
+            chi2s[indices] = np.nan
+            p[indices] = np.nan
```

Comment:
`i` appears unused; do you need the enumerate?

Reply:
Removed.